File: pserver/protocol/src/RespPackage.cpp
#include "stdafx.h"
#include "../inc/RespPackage.h"

RespPackage::RespPackage(ProtocolHeader header){
	fill(header);
}

RespPackage::RespPackage(ProtocolHeader header, char* body, int size){
	fill(header);
	fill(body, size);
}

void RespPackage::fill(ProtocolHeader header){
	this->header.mode = header.mode;
	this->header.seq = header.seq;
	for(int i = 0; i < ProtocolHeader::PIN_CNT; i++){
		this->header.pin[i] = header.pin[i];
	}
}

void RespPackage::fill(char* body, int size){
	this->header.length[0] = (char)(size & 0x000000FF);
	this->header.length[1] = (char)((size & 0x0000FF00) >> 8);
	this->body = body;
}

File: protocol/com/wireless/protocol/RespPackage.java
package com.wireless.protocol;

import com.wireless.protocol.parcel.Parcel;
import com.wireless.protocol.parcel.Parcelable;

public class RespPackage extends ProtocolPackage{
	
	public RespPackage(ProtocolHeader reqHeader){
		
		this.header.mode = reqHeader.mode;
		this.header.type = Type.ACK;
		this.header.seq = reqHeader.seq;
		System.arraycopy(reqHeader.pin, 0, header.pin, 0, header.pin.length);
		
		this.body = new byte[0];
	}
	
	public RespPackage(ProtocolHeader header, Parcel parcel){
		this(header);
		if(parcel != null){
			this.body = parcel.marshall();
		}else{
			this.body = new byte[0];
		}
	}
	
	public RespPackage(ProtocolHeader header, Parcelable parcelable, int flag){
		this(header);
		if(parcelable != null){
			Parcel p = new Parcel();
			parcelable.writeToParcel(p, flag);
			this.body = p.marshall();
			
		}else{
			this.body = new byte[0];
		}
	}
}
